### PR TITLE
Johnc-Nirvana

### PR DESCRIPTION
**Pull request by johnc1231**
Hopefully this is what you needed me to do. If not let me know. @jbloom22

### CODE REVIEW

**Comment by jbloom22** [comment]
I've rebased John's branch, added the default block_size change by @shulik7 at Nirvana, and made a few additional small changes including adding the @typecheck_method and @record_method decorators (the latter is now required, else history goes histrionic).

The configuration file, init script, and resource files are in`gs://hail-common/nirvana`.

The init script `gs://hail-common/nirvana/nirvana-init-GRCh37.sh` makes local copies of the resource files and .net: 
```
#!/bin/bash

mkdir -p /nirvana/Data/Cache
mkdir -p /nirvana/Data/References
mkdir -p /nirvana/Data/SupplementaryDatabase

#Data is copied for use with Nirvana 1.6.2 as of June 19 2017
gsutil -m cp -r gs://hail-common/nirvana/Data/Cache/24/GRCh37 /nirvana/Data/Cache
gsutil -m cp gs://hail-common/nirvana/Data/References/5/Homo_sapiens.GRCh37.Nirvana.dat /nirvana/Data/References
gsutil -m cp -r gs://hail-common/nirvana/Data/SupplementaryDatabase/39/GRCh37 /nirvana/Data/SupplementaryDatabase
gsutil -m cp -r gs://hail-common/nirvana/netcoreapp1.1 /nirvana
gsutil -m cp gs://hail-common/nirvana/nirvana-cloud-GRCh37.properties /nirvana

chmod -R 777 /nirvana

apt-get -y install curl libunwind8 gettext
curl -sSL -o dotnet.tar.gz https://go.microsoft.com/fwlink/?linkid=843453
mkdir -p /opt/dotnet && sudo tar zxf dotnet.tar.gz -C /opt/dotnet
ln -s /opt/dotnet/dotnet /usr/local/bin
```

The properties file `nirvana-cloud-GRCh37.properties` points Nirvana to these local resources:
```
hail.nirvana.location = /nirvana/netcoreapp1.1/Nirvana.dll
hail.nirvana.cache = /nirvana/Data/Cache/GRCh37/Ensembl84
hail.nirvana.reference = /nirvana/Data/References/Homo_sapiens.GRCh37.Nirvana.dat
hail.nirvana.supplementaryAnnotationDirectory = /nirvana/Data/SupplementaryDatabase/GRCh37
```

I started a cluster with the init script and ran Nirvana on all of `profile225.vcf`, and later exported results for just a region bounding the gene CABIN1:
```
from hail import *
hc = (HailContext())

(hc
 .import_vcf(path='gs://jbloom/profile225.vcf.bgz')
 .filter_multi()
 .nirvana(block_size=10000, config='/nirvana/nirvana-cloud-GRCh37.properties')
 .variants_table()
 .filter(expr='v.start > 24430000 && v.start < 24580000')
 .export(output='gs://jbloom/nirvana_cabin1.tsv'))
```

The top-level categories show reasonable-looking variation, except for "clinvar" and "genes" which are all `null` valued. Comparing a few variants in [gnomad](http://gnomad.broadinstitute.org/gene/ENSG00000099991), the annotations line up nicely.

Here's an example of a common missense variant:
```
22:24468386:G:A	{"rsid":null,"qual":38350.97,"filters":null,"info":{"AC":[306],"AF":[0.061],"AN":5018,"BaseQRankSum":26.807,"ClippingRankSum":-0.538,"DP":22432,"DS":null,"FS":1.203,"HaplotypeScore":null,"InbreedingCoeff":0.0335,"MLEAC":[309],"MLEAF":[0.062],"MQ":59.13,"MQ0":0,"MQRankSum":16.406,"QD":14.9,"ReadPosRankSum":-0.637,"set":null},"nirvana":{"chromosome":"22","refAllele":"G","position":24468386,"altAlleles":["A"],"cytogeneticBand":"22q11.23","filters":null,"variants":[{"altAllele":"A","refAllele":"G","chromosome":"22","begin":24468386,"end":24468386,"phylopScore":3.457,"isReferenceMinor":null,"variantType":"SNV","vid":"22:24468386:A","isRecomposed":null,"regulatoryRegions":null,"clinvar":null,"cosmic":[{"id":"COSM3759087","isAlleleSpecific":true,"refAllele":"G","altAllele":"A","gene":"CABIN1","sampleCount":2,"studies":[{"id":376,"histology":"carcinoma","primarySite":"large intestine"},{"id":null,"histology":"carcinoma","primarySite":"breast"}]}],"dbsnp":{"ids":["rs17854874"]},"evs":{"coverage":113,"sampleCount":6503,"allAf":0.052514,"afrAf":0.015887,"eurAf":0.071279},"exac":{"coverage":39,"allAf":0.062845,"allAc":7625,"allAn":121330,"afrAf":0.012409,"afrAc":129,"afrAn":10396,"amrAf":0.048409,"amrAc":560,"amrAn":11568,"easAf":0.156967,"easAc":1354,"easAn":8626,"finAf":0.051698,"finAc":341,"finAn":6596,"nfeAf":0.070106,"nfeAc":4678,"nfeAn":66728,"othAf":0.084071,"othAc":76,"othAn":904,"sasAf":0.029494,"sasAc":487,"sasAn":16512},"globalAllele":{"globalMinorAllele":"A","globalMinorAlleleFrequency":0.0613},"oneKg":{"ancestralAllele":"G","allAf":0.061302,"allAc":307,"allAn":5008,"afrAf":0.003782,"afrAc":5,"afrAn":1322,"amrAf":0.0317,"amrAc":22,"amrAn":694,"easAf":0.15873,"easAc":160,"easAn":1008,"eurAf":0.097416,"eurAc":98,"eurAn":1006,"sasAf":0.022495,"sasAc":22,"sasAn":978},"transcripts":{"refSeq":null,"ensembl":[{"transcript":"ENST00000263119.5","bioType":"protein_coding","aminoAcids":"R/Q","cDnaPos":"2685","codons":"cGg/cAg","cdsPos":"2558","exons":"18/37","introns":null,"geneId":"ENSG00000099991","hgnc":"CABIN1","consequence":["missense_variant"],"hgvsc":"ENST00000263119.5:c.2558G>A","hgvsp":"ENSP00000263119.5:p.(Arg853Gln)","isCanonical":null,"polyPhenScore":0.625,"polyPhenPrediction":"possibly damaging","proteinId":"ENSP00000263119.5","proteinPos":"853","siftScore":0.01,"siftPrediction":"deleterious"},{"transcript":"ENST00000405822.2","bioType":"protein_coding","aminoAcids":"R/Q","cDnaPos":"2502","codons":"cGg/cAg","cdsPos":"2408","exons":"17/36","introns":null,"geneId":"ENSG00000099991","hgnc":"CABIN1","consequence":["missense_variant"],"hgvsc":"ENST00000405822.2:c.2408G>A","hgvsp":"ENSP00000384694.2:p.(Arg803Gln)","isCanonical":null,"polyPhenScore":0.94,"polyPhenPrediction":"probably damaging","proteinId":"ENSP00000384694.2","proteinPos":"803","siftScore":0.02,"siftPrediction":"deleterious"},{"transcript":"ENST00000398319.2","bioType":"protein_coding","aminoAcids":"R/Q","cDnaPos":"2943","codons":"cGg/cAg","cdsPos":"2558","exons":"18/37","introns":null,"geneId":"ENSG00000099991","hgnc":"CABIN1","consequence":["missense_variant"],"hgvsc":"ENST00000398319.2:c.2558G>A","hgvsp":"ENSP00000381364.2:p.(Arg853Gln)","isCanonical":true,"polyPhenScore":0.625,"polyPhenPrediction":"possibly damaging","proteinId":"ENSP00000381364.2","proteinPos":"853","siftScore":0.01,"siftPrediction":"deleterious"},{"transcript":"ENST00000484593.1","bioType":"retained_intron","aminoAcids":null,"cDnaPos":null,"codons":null,"cdsPos":null,"exons":null,"introns":null,"geneId":"ENSG00000099991","hgnc":"CABIN1","consequence":["downstream_gene_variant"],"hgvsc":null,"hgvsp":null,"isCanonical":null,"polyPhenScore":null,"polyPhenPrediction":null,"proteinId":null,"proteinPos":null,"siftScore":null,"siftPrediction":null}]},"genes":null}]}}
```

I've added the experimental warning until there is some form of automated testing in place (perhaps validating that consistency is maintained against a fixed output file that's been otherwise reviewed by Nirvana).

File: python/hail/dataset.py [hunk 1]
```diff
@@ -5323,6 +5323,227 @@ def vep(self, config, block_size=1000, root='va.vep', csq=False):
...
+    @typecheck_method(config=strlike,
+                      block_size=integral,
+                      root=strlike)
+    def nirvana(self, config, block_size = 500000, root = 'va.nirvana'):
+        """Annotate variants using `Nirvana <https://github.com/Illumina/Nirvana>`_.
+        
+        .. include:: _templates/experimental.rst
+
+        ***Configuration***
+        :py:meth:`~hail.VariantDataset.nirvana` requires a configuration file. The format is a
```
Comment (tpoterba) [inline python/hail/dataset.py]: do we have one pre-written on the cloud?

File: python/hail/dataset.py [hunk 2]
```diff
@@ -5323,6 +5323,227 @@ def vep(self, config, block_size=1000, root='va.vep', csq=False):
...
+
+        Annotations with the following schema are placed in the location specified by ``root``.
+
+        .. code-block:: text
+        
+            Struct{
+                chromosome: String,
+                refAllele: String,
+                position: Int,
+                altAlleles: Array(String),
```
Comment (tpoterba) [inline python/hail/dataset.py]: Array[T], not Array(T)

File: src/main/scala/is/hail/methods/Nirvana.scala [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+import is.hail.variant.{Variant, VariantDataset}
+import org.apache.spark.storage.StorageLevel
+import org.json4s.jackson.JsonMethods
+
+import scala.collection.JavaConverters._
+
+
+object Nirvana {
+
+  //NOTE THIS SCHEMA IS FOR NIRVANA 1.6.2 as of JUNE 19th
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/Nirvana.scala]: wasn't the schema updated more recently?

File: src/main/scala/is/hail/methods/Nirvana.scala [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+          .grouped(localBlockSize)
+          .flatMap { block =>
+            val (jt, proc) = block.iterator.pipe(pb,
+              printContext,
+              printElement(oldSignature),
+              _ => ())
+            // The filter is because every other output line is a comma.
+            val kt = jt.filter(_.startsWith("{\"chromosome")).map { s =>
+                val a = JSONAnnotationImpex.importAnnotation(JsonMethods.parse(s), nirvanaSignature)
+                if(startQuery(a).asInstanceOf[Int] == 0) {
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/Nirvana.scala]: remove, or convert into assertion

**Comment by jbloom22** [comment]
made PR addressing comments against John's branch, he'll merge or give me write permission

File: python/hail/dataset.py [hunk 5]
```diff
@@ -5323,6 +5323,249 @@ def vep(self, config, block_size=1000, root='va.vep', csq=False):
 
     @handle_py4j
     @record_method
+    @typecheck_method(config=strlike,
+                      block_size=integral,
+                      root=strlike)
+    def nirvana(self, config, block_size = 500000, root = 'va.nirvana'):
```
Comment (tpoterba) [inline python/hail/dataset.py]: last change: PEP8 dictates no spaces around keyword args and default arguments, e.g. `root='va.nirvana'`.

**Pull request merged**